File: http.signature.ts
/*
  Copyright 2022 Loophole Labs

  Licensed under the Apache License, Version 2.0 (the "License");
  you may not use this file except in compliance with the License.
  You may obtain a copy of the License at

       http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
*/

import {
  encodeString,
  decodeString,
  Kind,
  encodeArray,
  decodeArray,
  encodeError,
  encodeInt64, encodeUint8Array, encodeMap, decodeInt64, decodeUint8Array, decodeMap, encodeInt32, decodeInt32
} from "@loopholelabs/polyglot-ts";

export class HttpContext {
  constructor(Request: HttpRequest, Response: HttpResponse) {
    this._Request = Request
    this._Response = Response
  }

  private _Request: HttpRequest;

  get Request(): HttpRequest {
    return this._Request
  }

  set Request(Request: HttpRequest) {
    this._Request = Request
  }

  private _Response: HttpResponse;

  get Response(): HttpResponse {
    return this._Response
  }

  set Response(Response: HttpResponse) {
    this._Response = Response
  }

  encode(buf: Uint8Array): Uint8Array {
    let encoded = buf
    encoded = this._Request.encode(encoded)
    encoded = this._Response.encode(encoded)
    return encoded
  }

  internalError(buf: Uint8Array, err: Error): Uint8Array {
    return encodeError(buf, err)
  }

  static decode(buf: Uint8Array): {
    buf: Uint8Array,
    value: HttpContext
  } {
    let decoded = buf
    const Req = HttpRequest.decode(decoded)
    decoded = Req.buf
    const Resp = HttpResponse.decode(decoded)
    decoded = Resp.buf
    return {
      buf: decoded,
      value: new HttpContext(Req.value, Resp.value)
    }
  }
}

export class HttpStringList {
  constructor(Value: string[]) {
    this._Value = Value
  }

  private _Value: string[];

  get Value(): string[] {
    return this._Value
  }

  set Value(Value: string[]) {
    this._Value = Value
  }

  encode(buf: Uint8Array): Uint8Array {
    let encoded = buf
    encoded = encodeArray(encoded, this._Value.length, Kind.String)
    this._Value.forEach(field => {
      encoded = encodeString(encoded, field);
    })
    return encoded
  }

  static decode(buf: Uint8Array): {
    buf: Uint8Array,
    value: HttpStringList
  } {
    let decoded = buf
    const ValueArray = decodeArray(decoded)
    decoded = ValueArray.buf
    const Value: { value: string[] } = {
      value: [],
    }

    for (let i = 0; i < ValueArray.size; i++) {
      const element = decodeString(decoded);
      decoded = element.buf;
      Value.value.push(element.value);
    }

    return {
      buf: decoded,
      value: new HttpStringList(Value.value)
    }
  }
}

export class HttpRequest {
  constructor(Method: string, URI: string, ContentLength: bigint, Protocol: string, IP: string, Body: Uint8Array, Headers: Map<string, HttpStringList>) {
    this._Method = Method
    this._URI = URI
    this._ContentLength = ContentLength
    this._Protocol = Protocol
    this._IP = IP
    this._Body = Body
    this._Headers = Headers
  }

  private _Method: string;

  get Method(): string {
    return this._Method
  }

  set Method(Method: string) {
    this._Method = Method
  }

  private _ContentLength: bigint;

  get ContentLength(): bigint {
    return this._ContentLength
  }

  set ContentLength(ContentLength: bigint) {
    this._ContentLength = ContentLength
  }

  private _Protocol: string;

  get Protocol(): string {
    return this._Protocol
  }

  set Protocol(Protocol: string) {
    this._Protocol = Protocol
  }

  private _IP: string;

  get IP(): string {
    return this._IP
  }

  set IP(IP: string) {
    this._IP = IP
  }

  private _Body: Uint8Array;

  get Body(): Uint8Array {
    return this._Body
  }

  set Body(Body: Uint8Array) {
    this._Body = Body
  }

  private _Headers: Map<string, HttpStringList>;

  get Headers(): Map<string, HttpStringList> {
    return this._Headers
  }

  set Headers(Headers: Map<string, HttpStringList>) {
    this._Headers = Headers
  }

  private _URI: string;

  get URI(): string {
      return this._URI
  }

  set URI(URI: string) {
      this._URI = URI
  }

  encode(buf: Uint8Array): Uint8Array {
    let encoded = buf
    encoded = encodeString(encoded, this._URI)
    encoded = encodeString(encoded, this._Method)
    encoded = encodeInt64(encoded, this._ContentLength)
    encoded = encodeString(encoded, this._Protocol)
    encoded = encodeString(encoded, this._IP)
    encoded = encodeUint8Array(encoded, this._Body)
    encoded = encodeMap(encoded, this._Headers.size,
        Kind.String, Kind.Any)
    this._Headers.forEach((value, key) => {
      encoded = encodeString(encoded, key);
      encoded = value.encode(encoded);
    })
    return encoded
  }

  static decode(buf: Uint8Array): {
    buf: Uint8Array,
    value: HttpRequest
  } {
    let decoded = buf
    const URI = decodeString(decoded)
    decoded = URI.buf
    const Method = decodeString(decoded)
    decoded = Method.buf
    const ContentLength = decodeInt64(decoded)
    decoded = ContentLength.buf
    const Protocol = decodeString(decoded)
    decoded = Protocol.buf
    const IP = decodeString(decoded)
    decoded = IP.buf
    const Body = decodeUint8Array(decoded)
    decoded = Body.buf
    const HeadersMap = decodeMap(decoded)
    decoded = HeadersMap.buf
    const Headers: { value: Map<string, HttpStringList> } = {
      value: new Map<string, HttpStringList>(),
    }

    for (let i = 0; i < HeadersMap.size; i++) {
      const key = decodeString(decoded);
      decoded = key.buf;
      const value = HttpStringList.decode(decoded);
      decoded = value.buf;
      Headers.value.set(key.value, value.value);
    }

    return {
      buf: decoded,
      value: new HttpRequest(Method.value, URI.value, ContentLength.value, Protocol.value, IP.value, Body.value, Headers.value)
    }
  }
}

export class HttpResponse {
  constructor(StatusCode: number, Body: Uint8Array, Headers: Map<string, HttpStringList>) {
    this._StatusCode = StatusCode
    this._Body = Body
    this._Headers = Headers
  }

  private _StatusCode: number;

  get StatusCode(): number {
    return this._StatusCode
  }

  set StatusCode(StatusCode: number) {
    this._StatusCode = StatusCode
  }

  private _Body: Uint8Array;

  get Body(): Uint8Array {
    return this._Body
  }

  set Body(Body: Uint8Array) {
    this._Body = Body
  }

  private _Headers: Map<string, HttpStringList>;

  get Headers(): Map<string, HttpStringList> {
    return this._Headers
  }

  set Headers(Headers: Map<string, HttpStringList>) {
    this._Headers = Headers
  }

  encode(buf: Uint8Array): Uint8Array {
    let encoded = buf
    encoded = encodeInt32(encoded, this._StatusCode)
    encoded = encodeUint8Array(encoded, this._Body)
    encoded = encodeMap(encoded, this._Headers.size,
        Kind.String, Kind.Any)
    this._Headers.forEach((value, key) => {
      encoded = encodeString(encoded, key);
      encoded = value.encode(encoded);
    })
    return encoded
  }

  static decode(buf: Uint8Array): {
    buf: Uint8Array,
    value: HttpResponse
  } {
    let decoded = buf
    const StatusCode = decodeInt32(decoded)
    decoded = StatusCode.buf
    const Body = decodeUint8Array(decoded)
    decoded = Body.buf
    const HeadersMap = decodeMap(decoded)
    decoded = HeadersMap.buf
    const Headers: { value: Map<string, HttpStringList> } = {
      value: new Map<string, HttpStringList>(),
    }

    for (let i = 0; i < HeadersMap.size; i++) {
      const key = decodeString(decoded);
      decoded = key.buf;
      const value = HttpStringList.decode(decoded);
      decoded = value.buf;
      Headers.value.set(key.value, value.value);
    }

    return {
      buf: decoded,
      value: new HttpResponse(StatusCode.value, Body.value, Headers.value)
    }
  }
}

File: http_signature.rs
/*
	Copyright 2022 Loophole Labs

	Licensed under the Apache License, Version 2.0 (the "License");
	you may not use this file except in compliance with the License.
	You may obtain a copy of the License at

		   http://www.apache.org/licenses/LICENSE-2.0

	Unless required by applicable law or agreed to in writing, software
	distributed under the License is distributed on an "AS IS" BASIS,
	WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
	See the License for the specific language governing permissions and
	limitations under the License.
*/

use polyglot_rs::{Decoder, DecodingError, Encoder, Kind};
use std::collections::HashMap;
use std::io::Cursor;

pub trait Encode {
    fn encode<'a>(&'a self, b: &'a mut Cursor<Vec<u8>>) -> Result<&mut Cursor<Vec<u8>>, Box<dyn std::error::Error>>;
    fn internal_error<'a>(&'a self, b: &'a mut Cursor<Vec<u8>>, error: Box<dyn std::error::Error>);
}

pub trait Decode {
    fn decode(b: &mut Cursor<&mut Vec<u8>>) -> Result<Option<Self>, Box<dyn std::error::Error>>
    where
        Self: Sized;
}

#[derive(Clone)]
pub struct HttpContext {
    pub(crate) request: HttpRequest,
    pub(crate) response: HttpResponse,
}

impl Encode for HttpContext {
    fn encode<'a>(&'a self, b: &'a mut Cursor<Vec<u8>>) -> Result<&mut Cursor<Vec<u8>>, Box<dyn std::error::Error>> {
        self.request.encode(b)?;
        self.response.encode(b)?;
        Ok(b)
    }

    fn internal_error<'a>(&'a self, b: &'a mut Cursor<Vec<u8>>, error: Box<dyn std::error::Error>) {
        b.encode_error(error).unwrap();
    }
}

impl Decode for HttpContext {
    fn decode(
        b: &mut Cursor<&mut Vec<u8>>,
    ) -> Result<Option<HttpContext>, Box<dyn std::error::Error>> {
        if b.decode_none() {
            return Ok(None);
        }

        if let Ok(error) = b.decode_error() {
            return Err(error);
        }

        Ok(Some(HttpContext {
            request: HttpRequest::decode(b)?.ok_or(DecodingError::InvalidStruct)?,
            response: HttpResponse::decode(b)?.ok_or(DecodingError::InvalidStruct)?,
        }))
    }
}

#[derive(Clone)]
pub struct HttpRequest {
    pub(crate) headers: HashMap<String, HttpStringList>,
    pub(crate) uri: String,
    pub(crate) method: String,
    pub(crate) content_length: i64,
    pub(crate) protocol: String,
    pub(crate) ip: String,
    pub(crate) body: Vec<u8>,
}

impl Encode for HttpRequest {
    fn encode<'a>(&'a self, b: &'a mut Cursor<Vec<u8>>) -> Result<&mut Cursor<Vec<u8>>, Box<dyn std::error::Error>> {
        b.encode_string(&*self.uri)?
            .encode_string(&*self.method)?
            .encode_i64(self.content_length)?
            .encode_string(&*self.protocol)?
            .encode_string(&*self.ip)?
            .encode_bytes(&self.body)?;

        b.encode_map(self.headers.len(), Kind::String, Kind::Any)?;
        for (k, v) in &self.headers {
            b.encode_string(&*k)?;
            v.encode(b)?;
        }
        Ok(b)
    }

    fn internal_error<'a>(&'a self, b: &'a mut Cursor<Vec<u8>>, error: Box<dyn std::error::Error>) {
        b.encode_error(error).unwrap();
    }
}

impl Decode for HttpRequest {
    fn decode(
        b: &mut Cursor<&mut Vec<u8>>,
    ) -> Result<Option<HttpRequest>, Box<dyn std::error::Error>> {
        if b.decode_none() {
            return Ok(None);
        }

        if let Ok(error) = b.decode_error() {
            return Err(error);
        }

        fn headers_decode(
            b: &mut Cursor<&mut Vec<u8>>,
        ) -> Result<Option<HashMap<String, HttpStringList>>, Box<dyn std::error::Error>> {
            if b.decode_none() {
                return Ok(None);
            }

            let size = b
                .decode_map(Kind::String, Kind::Any)
                .ok()
                .ok_or(DecodingError::InvalidU32)?;
            let mut map = HashMap::new();
            for _ in 0..size {
                let k = b.decode_string()?;
                let v = HttpStringList::decode(b)?.ok_or(DecodingError::InvalidMap)?;
                map.insert(k, v);
            }
            Ok(Some(map))
        }
        Ok(Some(HttpRequest {
            uri: b.decode_string()?,
            method: b.decode_string()?,
            content_length: b.decode_i64()?,
            protocol: b.decode_string()?,
            ip: b.decode_string()?,
            body: b.decode_bytes()?,
            headers: headers_decode(b)?.ok_or(DecodingError::InvalidMap)?,
        }))
    }
}

#[derive(Clone)]
pub struct HttpResponse {
    pub(crate) headers: HashMap<String, HttpStringList>,
    pub(crate) status_code: i32,
    pub(crate) body: Vec<u8>,
}

impl Encode for HttpResponse {
    fn encode<'a>(&'a self, b: &'a mut Cursor<Vec<u8>>) -> Result<&mut Cursor<Vec<u8>>, Box<dyn std::error::Error>> {
        b.encode_i32(self.status_code)?.encode_bytes(&self.body)?;

        b.encode_map(self.headers.len(), Kind::String, Kind::Any)?;
        for (k, v) in &self.headers {
            b.encode_string(&*k)?;
            v.encode(b)?;
        }
        Ok(b)
    }

    fn internal_error<'a>(&'a self, b: &'a mut Cursor<Vec<u8>>, error: Box<dyn std::error::Error>) {
        b.encode_error(error).unwrap();
    }
}

impl Decode for HttpResponse {
    fn decode(
        b: &mut Cursor<&mut Vec<u8>>,
    ) -> Result<Option<HttpResponse>, Box<dyn std::error::Error>> {
        if b.decode_none() {
            return Ok(None);
        }

        if let Ok(error) = b.decode_error() {
            return Err(error);
        }

        fn headers_decode(
            b: &mut Cursor<&mut Vec<u8>>,
        ) -> Result<Option<HashMap<String, HttpStringList>>, Box<dyn std::error::Error>> {
            if b.decode_none() {
                return Ok(None);
            }

            let size = b
                .decode_map(Kind::String, Kind::Any)
                .ok()
                .ok_or(DecodingError::InvalidU32)?;
            let mut map = HashMap::new();
            for _ in 0..size {
                let k = b.decode_string()?;
                let v = HttpStringList::decode(b)?.ok_or(DecodingError::InvalidMap)?;
                map.insert(k, v);
            }
            Ok(Some(map))
        }
        Ok(Some(HttpResponse {
            status_code: b.decode_i32()?,
            body: b.decode_bytes()?,
            headers: headers_decode(b)?.ok_or(DecodingError::InvalidMap)?,
        }))
    }
}

#[derive(Clone)]
pub struct HttpStringList {
    pub value: Vec<String>,
}

impl Encode for HttpStringList {
    fn encode<'a>(&'a self, b: &'a mut Cursor<Vec<u8>>) -> Result<&mut Cursor<Vec<u8>>, Box<dyn std::error::Error>> {
        b.encode_array(self.value.len(), Kind::String)?;
        for item in &self.value {
            b.encode_string(&*item)?;
        }
        Ok(b)
    }

    fn internal_error<'a>(&'a self, b: &'a mut Cursor<Vec<u8>>, error: Box<dyn std::error::Error>) {
        b.encode_error(error).unwrap();
    }
}

impl Decode for HttpStringList {
    fn decode(
        b: &mut Cursor<&mut Vec<u8>>,
    ) -> Result<Option<HttpStringList>, Box<dyn std::error::Error>> {
        if b.decode_none() {
            return Ok(None);
        }

        if let Ok(error) = b.decode_error() {
            return Err(error);
        }

        fn value_decode(
            b: &mut Cursor<&mut Vec<u8>>,
        ) -> Result<Option<Vec<String>>, Box<dyn std::error::Error>> {
            let value_size = b.decode_array(Kind::String)?;
            let mut temp = Vec::with_capacity(value_size);
            for _ in 0..value_size {
                temp.push(b.decode_string()?);
            }
            Ok(Some(temp))
        }
        Ok(Some(HttpStringList {
            value: value_decode(b)?.ok_or(DecodingError::InvalidArray)?,
        }))
    }
}
